diff --git a/src/build/GhosttyZig.zig b/src/build/GhosttyZig.zig
index 46ab05d..18b1605 100644
--- a/src/build/GhosttyZig.zig
+++ b/src/build/GhosttyZig.zig
@@ -123,7 +123,10 @@ fn initVt(
         // no-libcxx mode (HWY_NO_LIBCXX / SIMDUTF_NO_LIBCXX) so we
         // don't need libcpp. System-provided simdutf headers still
         // use C++ stdlib headers, so we need libcpp in that case.
-        .link_libc = if (cfg.simd) true else null,
+        // The Linux C embedder uses libc. Without it, Wuffs' hidden weak
+        // calloc/free stubs override glibc in the final ELF executable.
+        .link_libc = if (cfg.simd or (vt_options.c_abi and
+            cfg.target.result.os.tag == .linux)) true else null,
         .link_libcpp = if (cfg.simd and
             b.systemIntegrationOption("simdutf", .{}) and
             cfg.target.result.abi != .msvc) true else null,
